ring_doorbell/auth.py

```diff
@@ -47,44 +83,95 @@ def __fetch_token(self, username, password,
                 headers=headers)
 
-        return self._oauth.fetch_token(
+            self.__default_token_updater(token)
+
+            return token
+
+        token = self._oauth.fetch_token(
             OAuth.ENDPOINT,
             username=username,
             password=password,
-            scope=OAuth.SCOPE)
+            scope=OAuth.SCOPE,
+            headers=OAuth.HEADERS)
+
+        self.__default_token_updater(token)
+
+        return token
 
-    # Python 2 doesn't support typing for return value,
-    #  removed: -> Dict[str, Union[str, int]]:
     def refresh_tokens(self):
         """Refreshes the auth tokens"""
-        token = self._oauth.refresh_token(OAuth.ENDPOINT)
+        token = self._oauth.refresh_token(
+            OAuth.ENDPOINT, headers=OAuth.HEADERS)
 
         if self.token_updater is not None:
             self.token_updater(token)
 
+        self.__default_token_updater(token)
         return token
 
-    # Python 2 doesn't support typing for return value,
-    #  removed: -> Response:
-    def request(self, method, resource, **kwargs):
-        """Does an http request, if token is expired, then it will refresh
-        :type method: str
-        :type resource: str
-        """
-        try:
-            return self._oauth.request(method, resource, **kwargs)
+    def query(self,
+              url,
+              attempts=RETRY_TOKEN,
+              method='GET',
+              raw=False,
+              extra_params=None,
+              json=None,
+              timeout=None):
+        """Query data from Ring API."""
+        # Configure timeout specific to this query
+        query_timeout = timeout or TIMEOUT
+
+        response = None
+        loop = 0
+        while loop <= attempts:
+            # allow to override params when necessary
+            # and update self.params globally for the next connection
+            if extra_params:
+                params = self.params
+                params.update(extra_params)
+            else:
+                params = self.params
+
+            loop += 1
+            try:
+                if method == 'GET':
+                    req = self._oauth.get(
+                        url, params=urlencode(params),
+                        headers=OAuth.HEADERS,
+                        timeout=query_timeout)
+                elif method == 'PUT':
+                    req = self._oauth.put(
+                        url, params=urlencode(params),
+                        headers=OAuth.HEADERS,
+                        timeout=query_timeout)
+                elif method == 'POST':
+                    req = self._oauth.post(
+                        url, params=urlencode(params), json=json,
+                        headers=OAuth.HEADERS,
+                        timeout=query_timeout)
+            except TokenExpiredError:
+                self._oauth.token = self.refresh_tokens()
+                req = self.query(url, attempts, method, raw,
```

Comment:
Dangerous, this can start hammering the server with recursive calls. It's better to just call it again. See `request` example here https://developers.home-assistant.io/docs/en/api_lib_auth.html#sync-example-1